### PR TITLE
[SPARK-928][CORE] Add support for Unsafe-based serializer in Kryo

### PR DESCRIPTION
**Pull request by techaddict**
## What changes were proposed in this pull request?

Now since we have migrated to Kryo-3.0.0 in https://issues.apache.org/jira/browse/SPARK-11416, we can gives users option to use unsafe SerDer. It can turned by setting `spark.kryo.useUnsafe` to `true`
## How was this patch tested?

Ran existing tests

```
     Benchmark Kryo Unsafe vs safe Serialization: Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
      ------------------------------------------------------------------------------------------------
      basicTypes: Int unsafe:true                    160 /  178         98.5          10.1       1.0X
      basicTypes: Long unsafe:true                   210 /  218         74.9          13.4       0.8X
      basicTypes: Float unsafe:true                  203 /  213         77.5          12.9       0.8X
      basicTypes: Double unsafe:true                 226 /  235         69.5          14.4       0.7X
      Array: Int unsafe:true                        1087 / 1101         14.5          69.1       0.1X
      Array: Long unsafe:true                       2758 / 2844          5.7         175.4       0.1X
      Array: Float unsafe:true                      1511 / 1552         10.4          96.1       0.1X
      Array: Double unsafe:true                     2942 / 2972          5.3         187.0       0.1X
      Map of string->Double unsafe:true             2645 / 2739          5.9         168.2       0.1X
      basicTypes: Int unsafe:false                   211 /  218         74.7          13.4       0.8X
      basicTypes: Long unsafe:false                  247 /  253         63.6          15.7       0.6X
      basicTypes: Float unsafe:false                 211 /  216         74.5          13.4       0.8X
      basicTypes: Double unsafe:false                227 /  233         69.2          14.4       0.7X
      Array: Int unsafe:false                       3012 / 3032          5.2         191.5       0.1X
      Array: Long unsafe:false                      4463 / 4515          3.5         283.8       0.0X
      Array: Float unsafe:false                     2788 / 2868          5.6         177.2       0.1X
      Array: Double unsafe:false                    3558 / 3752          4.4         226.2       0.0X
      Map of string->Double unsafe:false            2806 / 2933          5.6         178.4       0.1X
```

### CODE REVIEW

**Comment by vectorijk** [comment]
@techaddict According to [this](https://github.com/EsotericSoftware/kryo#-disclaimer-about-using-unsafe-based-io-), I am just wondering what will happen when Unsafe-based IO is not compatible with Kryo's Input and Output streams.

**Comment by techaddict** [comment]
@vectorijk The class Kryo Serializer is not wire-compatible, so IMO we should be ok. Either one uses Unsafe or safe based at a time.
Comment from the KryoSerializer class
```Note that this serializer is not guaranteed to be wire-compatible across different versions of
  Spark. It is intended to be used to serialize/de-serialize data within a single
  Spark application.```

File: core/src/test/scala/org/apache/spark/scheduler/KryoBenchmark.scala
Comment (rxin) [inline core/src/test/scala/org/apache/spark/scheduler/KryoBenchmark.scala]: this benchmark is probably wrong. these numbers don't make sense.

File: core/src/test/scala/org/apache/spark/scheduler/KryoBenchmark.scala
Comment (rxin) [inline core/src/test/scala/org/apache/spark/scheduler/KryoBenchmark.scala]: i think the whole loop might've gotten dead code eliminated. also for loop is pretty expensive so you are probably benchmarking the wrong thing here.

**Comment by techaddict** [comment]
@rxin figured out the problem, was using `var N` and its value was changed before the benchmark ran so each benchmark was running effectvely for `N=5000`, thus giving an impression of dead code elimination.
Plus I was passing function to kryo instead of value in one of the cases, so that was also just wrong.

**Comment by techaddict** [comment]
@JoshRosen @rxin can you review this.

**Comment by techaddict** [comment]
ping @rxin @JoshRosen @andrewor14

**Comment by techaddict** [comment]
cc: @srowen @rxin @JoshRosen

**Comment by techaddict** [comment]
ping @JoshRosen

**Comment by maver1ck** [comment]
Up ?

File: core/src/test/scala/org/apache/spark/serializer/KryoSerializerSuite.scala
Comment (holdenk) [inline core/src/test/scala/org/apache/spark/serializer/KryoSerializerSuite.scala]: At first glance I don't see this function used anywhere I'm assuming this was meant to wrap some of the tests?

Comment (techaddict) [inline core/src/test/scala/org/apache/spark/serializer/KryoSerializerSuite.scala]: Yes will update the pr today.

**Comment by techaddict** [comment]
@holdenk Updated the PR, ready for review again.

**Comment by holdenk** [comment]
Excellent - lets see so @mateiz was the creator of this issue originally so we should check and see if he is available to review (although I suspect he is rather busy in general). If not we can maybe reach out to @JoshRosen and see if he has any review bandwidth? I'm not familiar enough with Kyro to do a meaningful review myself - but if everyone else is busy I can try and get up to speed on it and see what we can do while we wait for one of the committers to be available.

File: core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala
Comment (mateiz) [inline core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala]: Nit: I'd call this spark.kryo.unsafe to make it easier to remember

File: core/src/test/scala/org/apache/spark/serializer/KryoSerializerSuite.scala
Comment (mateiz) [inline core/src/test/scala/org/apache/spark/serializer/KryoSerializerSuite.scala]: Is there any way to refactor this into two Suites, where one extends from the other and has additional setup code (or a setup method that it overrides)? Otherwise we have to remember to add this construct to all tests.

**Review by mateiz — CHANGES_REQUESTED** [review]

**Comment by mateiz** [comment]
Looks pretty good overall! I made two small comments but it seems worthwhile to add in and it's not a huge change.

File: core/src/test/scala/org/apache/spark/serializer/KryoSerializerSuite.scala
Comment (mateiz) [inline core/src/test/scala/org/apache/spark/serializer/KryoSerializerSuite.scala]: (We have a few other examples suites where we just inherit from existing suites, such as SortShuffleSuite vs ShuffleSuite.)

**Comment by techaddict** [comment]
@mateiz updated the pr 👍

**Comment by mateiz** [comment]
@techaddict Cool, thanks! Just remembered a couple more things:
- Can you edit KryoSerializerSuite to set the flag to false? Otherwise we might silently end up with both suites testing on true if we ever change the default.
- Edit docs/configuration.md to add this new setting to the table about configs (near the other Kryo settings).

**Comment by techaddict** [comment]
@mateiz updated 👍

File: core/src/test/scala/org/apache/spark/serializer/UnsafeKryoSerializerSuite.scala
Comment (mateiz) [inline core/src/test/scala/org/apache/spark/serializer/UnsafeKryoSerializerSuite.scala]: Have you checked that this runs in the right order w.r.t. the code that initializes a SparkContext in KryoSerializerSuite / SharedSparkContext? You may want to print sc.conf in a test. Otherwise, it would be better to put conf.set("spark.kryo.unsafe", "true") at the top-level inside this class. SharedSparkContext does create a SparkContext in its beforeAll, but you seem to editing the conf after that is done.

Comment (techaddict) [inline core/src/test/scala/org/apache/spark/serializer/UnsafeKryoSerializerSuite.scala]: Ohh yes, fixed and tested 👍

File: core/src/test/scala/org/apache/spark/serializer/UnsafeKryoSerializerSuite.scala
Comment (rxin) [inline core/src/test/scala/org/apache/spark/serializer/UnsafeKryoSerializerSuite.scala]: don't you need to reset this back in afterAll?

File: core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala
Comment (rxin) [inline core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala]: we should document what this means.

File: core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala
Comment (rxin) [inline core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala]: one line here too

File: core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala
Comment (rxin) [inline core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala]:
```
private[spark] class KryoSerializerInstance(ks: KryoSerializer, useUnsafe: Boolean)
  extends SerializerInstance {
```

File: core/src/test/scala/org/apache/spark/serializer/KryoBenchmark.scala
Comment (rxin) [inline core/src/test/scala/org/apache/spark/serializer/KryoBenchmark.scala]: I find this function is over abstracted and actually makes it more confusing what's going on. I'd just remove it since it is almost no-op.

File: core/src/test/scala/org/apache/spark/serializer/KryoBenchmark.scala
Comment (rxin) [inline core/src/test/scala/org/apache/spark/serializer/KryoBenchmark.scala]: the use of call by name for fn is really confusing, especially the way you are calling it.

File: core/src/test/scala/org/apache/spark/serializer/KryoBenchmark.scala
Comment (rxin) [inline core/src/test/scala/org/apache/spark/serializer/KryoBenchmark.scala]: same thing here -- you are measuring the overhead of array fill and random number genreator a lot.

File: core/src/test/scala/org/apache/spark/serializer/KryoBenchmark.scala
Comment (rxin) [inline core/src/test/scala/org/apache/spark/serializer/KryoBenchmark.scala]: here too ...

**Comment by techaddict** [comment]
@rxin can you review again, all comments addressed 👍

**Comment by rxin** [comment]
LGTM pending Jenkins.

**Comment by rxin** [comment]
Merging in master. Thanks!

**Pull request closed without merging**
